hicommonwealth/commonwealth
# Session Keys PR2: Store session keys and sign data

## Merge Plan

* Review, merge, and deploy PR4 first (#2398), after testing on staging. This is a simpler PR that just focuses on adding in stubs for non-ETH chains.
* Review, merge, and deploy PR2 (this PR). This PR includes serialization and signing logic for non-ETH chains, mostly in `client/controllers/server/sessionSigners/*`, and also changes to the backend database and models for signing data with session keys.
  * This PR also adds a signing modal that handles re-signing data. To test this modal, you should clear localStorage to simulate a session key expiring.
  * We do not enforce checks on expired session keys at this time, and should review the performance of this PR in production before turning on those checks. A significant reason is that we may want to adjust those parameters to allow for long-lived sessions depending on usage patterns across different communities.

## Description

- Requests session signature, upon login and if sessions expire, using a new modal
- Stores session keys for each <chainBase, chainId> pair in localStorage, and retrieves and re-verifies them on page reload
- Uses session keys to sign actions corresponding to main Commonwealth actions (create/delete thread, comment, reaction)
- Stores `canvas_session`, `canvas_action`, and `canvas_hash` on backend for models

Test plan:
- Log in with one wallet
- Clear localStorage and refresh
- Go to react to a comment or thread, and/or comment or post a thread. You should see a prompt to re-sign at the moment of creating your new content item
- Repeat with other wallets and chains
- Inspect the database, note that canvas_action, canvas_session, canvas_hash are all filled in

## Motivation and Context

Supporting decentralized data across all chains.

## How has this been tested?

For each of these communities, login using all wallets, create a thread, create a comment, react to the comment. Then create a second thread, create a comment, delete the comment, delete the thread.

- [x] Ethereum: http://localhost:8080/dydx
- [x] Evmos / Injective: http://localhost:8080/evmos
- [x] Cosmos: http://localhost:8080/osmosis
- [x] NEAR: http://localhost:8080/near
- [x] Solana: http://localhost:8080/solana
- [x] Polkadot: http://localhost:8080/edgeware

## Does this PR affect any server routes?

## If this PR affects server routes, what are the security implications?

n/a - no logical changes to server routes, other than new fields added.

The sign in format for verifyAddress has been updated; worth noting for non-Eth chains.

## Have proper tags been added (for bug, enhancement, breaking change)?
